File: py/data_structures/BST.py
# 二叉查找树


# 获取当前节点左子树中的最大值, 即访问最右侧节点
def get_max_left(node):
    cur_node = node
    while cur_node.right:
        cur_node = cur_node.right
    return cur_node


class BST:

    # 最终遍历之后展示的数组
    # 这个玩意如果写在这里，每次计算都会保留上一次的结果
    # order_list = []

    def __init__(self):
        # 根节点
        self.root = None
        self.order_list = []

    # 插入节点
    def insert(self, data):
        cur_node = self.root
        insert_node = BNode(data)
        if cur_node is None:
            self.root = insert_node
        else:
            # 遍历节点，将小的元素放到左边，大的放到右边
            while True:
                parent_node = cur_node
                if insert_node.data <= cur_node.data:
                    cur_node = cur_node.left
                    if cur_node is None:
                        parent_node.left = insert_node
                        break
                else:
                    cur_node = cur_node.right
                    if cur_node is None:
                        parent_node.right = insert_node
                        break

    # 先序遍历
    def pre_order(self, node):
        if node is not None:
            self.order_list.append(node.data)
            self.pre_order(node.left)
            self.pre_order(node.right)

    # 中序遍历
    def mid_order(self, node):
        # 遍历前重置元素序列
        # if node == self.root:
        #     self.order_list = []
        if node is not None:
            self.mid_order(node.left)
            self.order_list.append(node.data)
            self.mid_order(node.right)

    # 后序遍历
    def post_order(self, node):
        if node is not None:
            self.post_order(node.left)
            self.post_order(node.right)
            self.order_list.append(node.data)

    # 查找最小值, 找到最左边的元素即可
    def find_min(self):
        node = self.root
        while node.left:
            node = node.left
        return node.data

    # 查找最大值, 返回最右侧的元素
    def find_max(self):
        node = self.root
        while node.right:
            node = node.right
        return node.data

    # 查找元素
    def find(self, data):
        cur_node = self.root
        while cur_node:
            if cur_node.data == data:
                return True
            else:
                if data < cur_node.data:
                    cur_node = cur_node.left
                else:
                    cur_node = cur_node.right
        return False

    # 删除节点
    def remove(self, data):
        root = self.__remove_node(self.root, data)
        return root

    # 根据节点内容, 处理待删除子树剩余节点
    def __remove_node(self, node, data):
        if node is None:
            return None
        else:
            # 当前节点是待删除节点
            if node.data == data:
                # 待删除节点没有子节点
                if node.left is None and node.right is None:
                    return None
                # 无左子树有右子树
                elif node.left is None:
                    node = node.right
                    return node
                # 无右子树有左子树
                elif node.right is None:
                    node = node.left
                    return node
                # 左右子树均存在
                # 这里取左子树中的最大值替换原来的节点
                # -- 另一种方案是取右子树中的最小值，已经在js版本中实现
                else:
                    tem_node = get_max_left(node.left)
                    node.data = tem_node.data
                    node.left = self.__remove_node(node.left, tem_node.data)
                    return node

            # 数据比当前节点小，向左递归
            elif data < node.data:
                node = self.__remove_node(node.left, data)
                return node
            # 数据比当前节点大，向右递归
            else:
                node = self.__remove_node(node.right, data)
                return node


# 树节点
class BNode:
    def __init__(self, data):
        self.data = data
        # 左子树
        self.left = None
        # 右子树
        self.right = None

File: js/data_structures/BST.js
/**
 * binary search tree
 * 二叉查找树
 * 重点条件：
 * 左子树中的所有节点的数据域都小于或等于根节点的数据域，
 * 而右子树中的所有节点的数据域都大于等于根节点的数据域
 */
class BST {
    constructor() {
        // 根节点
        this.root = null
        this.dataList = []; // 遍历节点的数组
    }
    // bst的插入方法
    insert(data) {
        const insertNode = new BNode(data)
        // 根节点插入
        if (!this.root) {
            return this.root = insertNode; 
        }
        let curNode = this.root;
        let parent = null;
        while(true) {
            parent = curNode
            if (data <= curNode.data) {
                // 存入数据比当前节点数据小，向左遍历
                curNode = curNode.left
                if (curNode == null) { // 找到空位置, 插入
                    return parent.left = insertNode
                }
            } else {
                // 存入数据比当前节点数据大，向右遍历
                curNode = curNode.right
                if (curNode == null) {
                    return parent.right = insertNode
                }
            }
        }
    }
    /**
     * 中序遍历, 使用递归实现
     * 先遍历左子树，再到根节点，再到右子树
     */
    midOrder(node) {
        if (node != null) {
            this.midOrder(node.left);
            this.dataList.push(node.data)
            this.midOrder(node.right)
        }
    }
    /**
     * 先序遍历
     * 根 -> 左子树 -> 右子树
     */
    preOrder(node) {
        if (node) {
            this.dataList.push(node.data)
            this.preOrder(node.left)
            this.preOrder(node.right)
        }
    }
    /**
     * 后序遍历
     * 左子树 -> 右子树 -> 根
     */
    postOrder(node) {
        if (node) {
            this.postOrder(node.left)
            this.postOrder(node.right)
            this.dataList.push(node.data)
        }
    }
    // 查找最小值, 最左边节点
    findMin() {
        let curNode = this.root;
        while(curNode.left) {
            curNode = curNode.left
        }
        return curNode.data
    }
    // 顺着当前node, 查找最小值的节点
    findMinNode(node) {
        let curNode = node;
        while (curNode.left) {
            curNode = curNode.left
        }
        return curNode
    }
    // 查找最大值, 最右边节点
    findMax() {
        let curNode = this.root;
        while(curNode.right) {
            curNode = curNode.right
        }
        return curNode.data
    }
    // 查找特定的值
    find(data) {
        let curNode = this.root;
        while (curNode) {
            if (curNode.data == data) {
                return curNode;
            } else {
                if (data < curNode.data) {
                    curNode = curNode.left
                } else {
                    curNode = curNode.right
                }
            }
        }
        return null
    }
    /**
     * 二叉树删除节点操作
     * 仍然使用递归进行,
     * 1. 找到待删除节点
     * 2. 判断待删除节点是否有子节点
     * 3. 单个子节点时，挪上去
     * 4. 两个子节点时，可选左边上去或者右边上去
     */
    remove(data) {
        const root = this.removeNode(this.root, data)
        return root
    }
    removeNode(node, data) {
        if (!node) {
            return null
        }
        // 找到该节点
        if (node.data == data) {
            // 没有子节点
            if (!node.left && !node.right) {
                return null
            } else if (!node.left) { // 只有右节点
                return node.right
            } else if (!node.right) { // 只有左节点
                return node.left
            } else {
                /**
                 * 有两个节点
                 * 找到最右边相对于需要删除的节点的最小值
                 * 或者找到最左边相对于需要删除节点的最大值
                 */
                const temNode = this.findMinNode(node.right)
                node.data = temNode.data;
                node.right = this.removeNode(node.right, temNode.data)
                return node

            }
        } else if (data < node.data) {
            node.left = this.removeNode(node.left, data)
            return node
        } else {
            node.right = this.removeNode(node.right, data)
            return node
        }
    }
    show() {
        return this.dataList
    }
}

/**
 * 二叉树节点
 * 左子树，右子树
 */
class BNode {
    constructor(data) {
        this.data = data
        this.left = null;
        this.right = null;
    }
    show() {
        return this.data
    }
}

module.exports = BST